src/shared/tpm2-util.c

```diff
@@ -476,6 +476,59 @@ void tpm2_pcr_mask_to_selection(uint32_t mask, uint16_t bank, TPML_PCR_SELECTION
         };
 }
 
+static void tpm2_log_debug_buffer(const void *buffer, const size_t size, const char *msg) {
+        if (!DEBUG_LOGGING || !buffer || !size)
+                return;
+
+        _cleanup_free_ char *h = hexmem(buffer, size);
+        if (h)
+                log_debug("%s: %s", msg ?: "Buffer", h);
+        else
+                log_oom();
+}
+
+static void tpm2_log_debug_digest(const TPM2B_DIGEST *digest, const char *msg) {
+        if (digest)
+                tpm2_log_debug_buffer(digest->buffer, digest->size, msg ?: "Digest");
+}
+
+static int tpm2_get_policy_digest(
+                Tpm2Context *c,
+                const Tpm2Handle *session,
+                TPM2B_DIGEST **ret_policy_digest) {
+
+        _cleanup_(Esys_Freep) TPM2B_DIGEST *policy_digest = NULL;
+        TSS2_RC rc;
+
+        assert(c);
+        assert(session);
+
+        if (DEBUG_LOGGING || ret_policy_digest) {
+                log_debug("Acquiring policy digest.");
+
+                rc = sym_Esys_PolicyGetDigest(
+                                c->esys_context,
+                                session->esys_handle,
+                                ESYS_TR_NONE,
+                                ESYS_TR_NONE,
+                                ESYS_TR_NONE,
+                                &policy_digest);
+                if (rc != TSS2_RC_SUCCESS)
+                        return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                               "Failed to get policy digest from TPM: %s", sym_Tss2_RC_Decode(rc));
+
+                tpm2_log_debug_digest(policy_digest, "Session policy digest");
+        }
+
+        if (ret_policy_digest) {
+                /* Free any previous digest before assignment */
+                Esys_Freep(ret_policy_digest);
```

Comment:
please don't. we have this rule in systemd that output parameters should be output parameters only, and are fully initialized on success and not touched on failure. Here, you make it an input *and* output parameter though (which is OK in some special cases, but i don't think this qualifies). In particular, an input/output param should not be called `ret_`, because that's what identifies it as output param.

but anyway, please just drop this line, let the caller destroy the old value, not this function.

Reply:
will do.